src/main/java/graphql/nadel/engine/AsyncIsFieldAllowed.java

```diff
@@ -0,0 +1,73 @@
+package graphql.nadel.engine;
+
+import graphql.GraphQLError;
+import graphql.language.Field;
+import graphql.language.Node;
+import graphql.nadel.engine.transformation.OverallTypeInfo;
+import graphql.nadel.engine.transformation.OverallTypeInformation;
+import graphql.nadel.hooks.ServiceExecutionHooks;
+import graphql.schema.GraphQLFieldDefinition;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static graphql.introspection.Introspection.TypeNameMetaFieldDef;
+import static graphql.nadel.dsl.NodeId.getId;
+
+public class AsyncIsFieldAllowed {
+
+    private Map<String, GraphQLError> fieldIdsToErrors = new ConcurrentHashMap<>();
+    private final ServiceExecutionHooks serviceExecutionHooks;
+    private final NadelContext nadelContext;
+    private final OverallTypeInformation<?> overallTypeInformation;
+
+    public AsyncIsFieldAllowed(ServiceExecutionHooks serviceExecutionHooks, NadelContext nadelContext, OverallTypeInformation<?> overallTypeInformation) {
+        this.serviceExecutionHooks = serviceExecutionHooks;
+        this.nadelContext = nadelContext;
+        this.overallTypeInformation = overallTypeInformation;
+    }
+
+    public CompletableFuture<Map<String, GraphQLError>> isFieldAllowed(Node<?> root) {
+        return visitChildren(root).thenApply(unused -> {
```

Comment:
Does this ever visit the root node? Doesn't that mean that we always bypass the check on the top level field?